otlpjsonfilereceiver: add a replay_file config option to support heartbeats

ridwanmsharif opened the pull request:
**Description:** <Describe what has changed.>
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
This change adds a config option to allow users to replay static telemetry at an interval (poll_interval is the configurable interval).

This is useful for some  usecases like:
- Sending a heartbeat log to a logging backend, as a signal that the logging agent is alive
- Sending static metadata as metrics to metrics backend. The metadata can be read using the `otlpjsonfilereceiver` that uses this package and config.

**Link to tracking Issue:** 
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/31533

**Testing:**
- Manual testing
- unit test for the fileconsumer
- test for the `otlpjsonfilereceiver`

**Documentation:** <Describe the documentation added.>

Commit:
pkg/stanza: add a replay_file config option to support heartbeats

This change adds a config option to allow users to replay static
telemetry at an interval (poll_interval is the configurable interval).

This is useful for some  usecases like:
- Sending a heartbeat log to a logging backend, as a signal that the
  logging agent is alive
- Sending static metadata as metrics to metrics backend. The metadata
  can be read using the `otlpjsonfilereceiver` that uses this package
  and config.

Signed-off-by: Ridwan Sharif <ridwanmsharif@google.com>

atoulme commented:
Please add a changelog. Do you mean to use the PollInterval setting to set the replay interval?

ridwanmsharif commented:
> Please add a changelog. Do you mean to use the PollInterval setting to set the replay interval?

Yes, `PollInterval` is what you would use to configure the intervals for the replay

djaglowski commented on pkg/stanza/fileconsumer/config.go:
```diff
@@ -80,6 +80,7 @@ type Config struct {
 	FlushPeriod        time.Duration   `mapstructure:"force_flush_period,omitempty"`
 	Header             *HeaderConfig   `mapstructure:"header,omitempty"`
 	DeleteAfterRead    bool            `mapstructure:"delete_after_read,omitempty"`
+	ReplayFile         bool            `mapstructure:"replay_file,omitempty"`
```
I'd like to keep this flag out of the user facing API for the package, since it's not clear that it's broadly applicable.

What I'd like to do is put an options pattern in place, which will help with several other changes I anticipate for this package. Give me a day or two to work out what this looks like and then we can rebase to make this an option.

djaglowski replied:
I opened https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31596 to establish an options pattern. Once that is merged, let's add a `WithNoTracking` option. (For context on the suggested name, see [#31506](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31506#pullrequestreview-1917375290).) Then we can just pass in this option from only otlpjson receiver, [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/f27143fae73a4a0631db8c55e46f6821d6121f76/receiver/otlpjsonfilereceiver/file.go#L76).

ridwanmsharif replied:
I agree with this pattern here, I think this replay option is only useful for the `otlpjsonfilereceiver`. Thanks for being so quick

djaglowski replied:
#31596 has been merged

ridwanmsharif replied:
d87a14c6ab4c7c80aa41b04c50a1bcb219315995 makes it so that when `noTracking` is set using the option, no known or previous files are tracked. Wasn't sure if we should wait for #31506 to land so the manager doesn't have to make these `!noTracking` checks everywhere. But I'm fine either way

djaglowski replied:
Good point. Might be easier overall to wait for #31506

ridwanmsharif replied:
Sweet, I'll update this PR today or tomorrow!

ridwanmsharif replied:
I lied earlier, i got distracted instead. I did update it now, PTAL

Commit:
pkg/stanza: remove replay_config option and use WithNoTracking option instead

This change does the following:
- Removes the replay_file option from the filecononsumer config, in
  favor of a build option
- When no tracking is used, we don't keep track of known or previously
  polled files. Instead reading only currently polled files that fit the
  pattern
- Adds tests

Commit:
fileconsumer: skip all file tracking in memory if noTracking is enabled

Signed-off-by: Ridwan Sharif <ridwanmsharif@google.com>

djaglowski reviewed (commented):
Nice, I like how this came out. It gives me hope we can move further in the direction of modularizing various aspects of the fileconsumer package.

djaglowski commented:
Code looks good to me generally, just some CI failures to address

djaglowski reviewed (approved):
Thanks @ridwanmsharif!
